Test/NumberTheory/FastPower.test.cpp

```diff
@@ -0,0 +1,28 @@
+#define FAST_POWER_TEST
+#define CATCH_CONFIG_MAIN
+
+#include "../catch.hpp"
+#include "../../NumberTheory/FastPower.cpp"
+
+
+TEST_CASE("Base cases", "[fast_power]") {
+    //REQUIRE(fast_power(0,0) == undefined);
+    REQUIRE(fast_power(0,1) == 0);
+    REQUIRE(fast_power(1,0) == 1);
+    REQUIRE(fast_power(1,1) == 1);
+}
+
+TEST_CASE("Normal cases", "[fast_power]") {
+    REQUIRE(fast_power(2,2) == 4);
+    REQUIRE(fast_power(2,4) == 16);
+    REQUIRE(fast_power(3,4) == 81);
+    REQUIRE(fast_power(7,9) == 40353607);
+    REQUIRE(fast_power(15,10) == 576650390625);
+}
+
+TEST_CASE("Overflow cases", "[fast_power]") {
+    REQUIRE(fast_power(2,100) == 68719476736);
+    REQUIRE(fast_power(10,99) == 4440381706574496940U);
```

Comment:
Both these tests are failing, and I think the reason is that at some step of the calculation, the multiplication produces the value 2<sup>64</sup>, which for a 64-bit unsigned integer (`unsigned long long`) is 0, and so the final value also becomes 0 (as anything multiplied by 0 is 0).
This _could_ be solved by taking the modulus of the values at each step, like you suggested previously, but the value of that (prime) number should be as close as possible to 2<sup>64</sup>-1, to get exact results for a larger range of values.